dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/statelisteners/ElasticListener.java

```diff
@@ -98,15 +100,35 @@ private void eagerLoadEntry(Entry entry) {
         Hibernate.initialize(entry.getAliases());
     }
 
+    private String determineIndex(Entry entry) {
+        if (entry instanceof Tool || entry instanceof AppTool) {
+            return TOOLS_INDEX;
+        }
+        if (entry instanceof Notebook) {
+            return NOTEBOOKS_INDEX;
+        }
+        if (entry instanceof BioWorkflow) {
+            return WORKFLOWS_INDEX;
+        }
+        // Services are not currently indexed, see #2771
+        return null;
+    }
+
+    private List<Entry> filterEntriesByIndex(List<Entry> entries, String index) {
+        return entries.stream().filter(entry -> Objects.equals(determineIndex(entry), index)).toList();
+    }
+
     @Override
     public void handleIndexUpdate(Entry entry, StateManagerMode command) {
         eagerLoadEntry(entry);
         entry = filterCheckerWorkflows(entry);
-        // #2771 will need to disable this and properly create objects to get services into the index
-        entry = entry instanceof Service ? null : entry;
         if (entry == null) {
```

Comment:
You technically don't need this since `determineIndex` will handle a null entry, but maybe it's better for clarity.